trie.h

```diff
@@ -0,0 +1,37 @@
+#ifndef TRIE_H
+#define TRIE_H
+
+#include <map>
+#include <vector>
+#include <string>
+#include <memory>
+
+using std::string;
+using std::map;
+using std::vector;
+using std::shared_ptr;
+
+class Trie {
+private:
+    struct Node {
+        map<char, shared_ptr<Node>> children;
+        bool value;
+    };
+
+    shared_ptr<Node> head;
+
```

Comment:
STL is header only which is actually fine. I'd rather have everything bundled in the header, or a clear separation.

The thing that bugs me about pimple is the extra typing because you are basically manually doing virtual function calls yourself. So the latter case is probably better. Have a 'static Trie* Trie::Create()' function that creates the TrieImpl class (or whatever you want to call it in trie.cpp). So the Trie header is nice and clean minus that one weird static function.